### ISSUE TITLE
functions_categories:  TOPMOST_CATEGORY_PARENT_ID type

### ISSUE THREAD

**Issue by torvista** [zencart/zencart#6823]
In multiple places this is defined in the docblocks as int when it is a string.
Assuming that the end-game is that the docblocks match the type hinting and of course the usage, how best to square these up for this constant?

I've already done a preliminary review of this file adding type hinting and return types for the obviously correct, and not breaking anything, but leaving many doubts for individual resolution, such as this.

**Comment by drbyte** [comment]
While yes it's a string because it comes from the database as a string, the "string"'s content is always expected to be an integer (never float, never alpha, never symbols).

What's the problem you're trying to fix?

**Comment by torvista** [comment]
For example four instances of being declared as int, so if it was type hinted as such it would fail.

https://github.com/zencart/zencart/blob/3696a52b92360b9dee1eb9010936ec46930a56fa/includes/functions/functions_categories.php#L135-L140

https://github.com/zencart/zencart/blob/3696a52b92360b9dee1eb9010936ec46930a56fa/includes/functions/functions_categories.php#L183-L185

**Comment by drbyte** [comment]
Correct. That's why it's not typehinted.

The docblock says `int` because we intend to use that parameter as an integer value. 
And it's not typehinted as such because we know it often arrives as a string when called from other parts of the application, and thus cannot strictly enforce it.
Then, inside the function we commonly cast it to int when using it, or do other things with it if it is found to not be an integer value.

**Comment by torvista** [comment]
I was going to say should not the docblock match the parameter hint, but then I read around the thing a bit more and there are arguments for only adding docblock info if there is no type hinting (which should be done where possible) or if the type hinting needs more explanation....but if the usage is obvious, don't repeat the info.

Perhaps some policy should be defined to prevent blindly following IDE recomendations.

**Comment by drbyte** [comment]
Unofficial policy:
- typehints are preferable, and no need for docblock if all parameters are typehinted
- if a typehint cannot be used, then a docblock is desireable if the parameter's variable name isn't self-explanatory
- if documentation is required, then docblock is the place for it
- if a variable name could be safely renamed to avoid the need for documentation, that is preferable because it is more humanly intuitive

**Comment by torvista** [comment]
If a parameter cannot be type-hinted, is that not an indication that the various calling code chunks should be reviewed/rationalised to pass the same type? Where possible.

**Comment by drbyte** [comment]
> If a parameter cannot be type-hinted, is that not an indication that the various calling code chunks should be reviewed/rationalised to pass the same type? Where possible.

Yes. Or the entire application be rearchitected to not treat all database responses and user-inputs as strings without translating them into specific data types before they're ever used elsewhere in the application.
But end-users, and store-modders, will likely find such changes crippling to their code-customization desires. Not to mention upgraders.
